Client/Frontend/Browser/Favorites/FavoritesViewController.swift

```diff
@@ -742,7 +516,226 @@ extension FavoritesViewController {
         } else {
             // User doesn't want to see the recent searches option again
             Preferences.Search.shouldShowRecentSearchesOptIn.value = false
-            collectionView.reloadData()
         }
     }
 }
+
+// MARK: - Diffable data source + NSFetchedResultsControllerDelegate
+extension FavoritesViewController: NSFetchedResultsControllerDelegate {
+    
+    private var recentSearchesSectionExists: Bool {
+        availableSections[safe: Section.recentSearches.rawValue] != nil
+    }
+    
+    /// Performs first frc fetches and handles setting first snaphot.
+    private func initialSnapshotSetup() {
+        do {
+            try favoritesFRC.performFetch()
+        } catch {
+            log.error("Favorites fetch error: \(String(describing: error))")
+        }
+        
+        fetchRecentSearches()
+        
+        var snapshot = Snapshot()
+        snapshot.appendSections(availableSections)
+        
+        let favorites: [DataWrapper] = favoritesFRC.fetchedObjects?.compactMap {
+            return .favorite(FavoriteDiffable(objectID: $0.objectID, title: $0.displayTitle, url: $0.url))
+        } ?? []
+        
+        snapshot.appendItems(favorites, toSection: .favorites)
+        
+        if recentSearchesSectionExists, let objects = recentSearchesFRC.fetchedObjects {
+            snapshot.appendItems(objects.compactMap({ .recentSearch($0.objectID) }),
+                                 toSection: .recentSearches)
+        }
+        
+        dataSource.apply(snapshot, animatingDifferences: false)
+    }
+    
+    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>,
+                    didChangeContentWith snapshot: NSDiffableDataSourceSnapshotReference) {
+        
+        // Skip applying snapshots before first `initialSnapshotSetup` is made.
+        if dataSource.snapshot().numberOfSections == 0 {
+            return
+        }
+        
+        let currentSnapshot = dataSource.snapshot()
+        var newSnapshot = Snapshot()
+        newSnapshot.appendSections(availableSections)
+        
+        guard let ids = snapshot.itemIdentifiers as? [NSManagedObjectID] else { return }
+        
+        if controller === favoritesFRC {
+            var items = [DataWrapper]()
```

Comment:
I think this is fine as-is, but there's a possibility to break out each of the body for these `if controller === <condition>` into a separate method (ex: `onUpdateFavorites`, or `onUpdateRecentSearches`), so that you can easily unit test. For example, you can have unit tests which checks that when `recentSearchesSectionExists` is true, it appends items to `newSnapshot` and that nothing happens if `recentSearchesSectionExists` is false

As the code becomes more complex, this can help catch regressions. And when new bugs are discovered, you can create a test checking for that specific bug (to make sure it doesn't happen again). Just some food for thought 😄